src/model.rs

```diff
@@ -0,0 +1,768 @@
+                    let new_col_coords = right_most_col_coords.iter().map(|c| {
+                        (c.row(), NonZeroU32::new(c.col().get() + 1).unwrap())
+                    });
+
+                    let parent = coord.parent().unwrap();
+                    if let Some(Grammar{ kind: Kind::Grid(sub_coords), name, style }) = self.grammars.get(&parent) {
+                        let mut new_sub_coords = sub_coords.clone();
+                        let mut grammars = self.grammars.clone();
+                        for c in new_col_coords {
+                            grammars.insert(Coordinate::child_of(&parent.clone(), c), Grammar::default());
+                            new_sub_coords.push(c);
+                        }
+                        grammars.insert(parent, Grammar {
+                            kind: Kind::Grid(new_sub_coords.clone()),
+                            name: name.clone(),
+                            style: style.clone()
+                        });
+                        self.grammars = grammars;
+                    }
+                }
+                true
+            }
+            Action::InsertRow => {
+                if let Some(coord) = self.active_cell.clone() {
+                    // find the bottom-most coord
+                    let mut bottom_most_coord = coord.clone();
+                    while let Some(below_coord) = bottom_most_coord.neighbor_below() {
+                        if self.grammars.contains_key(&below_coord) {
+                            bottom_most_coord = below_coord;
+                        } else { break }
+                    }
+
+                    let bottom_most_row_coords = self.query_row(bottom_most_coord.full_row());
+                    let new_row_coords = bottom_most_row_coords.iter().map(|c| {
+                        (NonZeroU32::new(c.row().get() + 1).unwrap(), c.col())
+                    });
+
+                    let parent = coord.parent().unwrap();
+                    if let Some(Grammar{ kind: Kind::Grid(sub_coords), name, style }) = self.grammars.get(&parent) {
+                        let mut new_sub_coords = sub_coords.clone();
+                        let mut grammars = self.grammars.clone();
+                        for c in new_row_coords {
+                            grammars.insert(Coordinate::child_of(&parent.clone(), c), Grammar::default());
+                            new_sub_coords.push(c);
+                        }
+                        grammars.insert(parent, Grammar {
+                            kind: Kind::Grid(new_sub_coords.clone()),
+                            name: name.clone(),
+                            style: style.clone()
+                        });
+                        self.grammars = grammars;
+                    }
+                }
+                true
+            }
+        }
+    }
+
+    fn view(&self) -> Html {
```

Comment:
maybe we should also move these `view_` functions out into a new view.rs file ? would cut down the bloat and help us keep all `html!`-related code in one place

Reply:
good idea, will do.